lib/internal/streams/from.js

```diff
@@ -75,23 +79,29 @@ function from(Readable, iterable, opts) {
   }
 
   async function next() {
-    try {
-      const { value, done } = await iterator.next();
-      if (done) {
-        readable.push(null);
-      } else {
-        const res = await value;
-        if (res === null) {
-          reading = false;
-          throw new ERR_STREAM_NULL_VALUES();
-        } else if (readable.push(res)) {
-          next();
+    for (;;) {
+      try {
+        const { value, done } = isAsync ?
+          await iterator.next() :
+          iterator.next();
+
+        if (done) {
+          readable.push(null);
         } else {
-          reading = false;
+          const res = PromiseResolve(value) === value ? await value : value;
```

Comment:
@benjamingr Is this the best way to check if a value is a promise or not?

Reply:
Is the value guaranteed to be either an ES Promise object or a plain value? Couldn't it be a thenable for example?

Reply:
Yeah that check is incorrect - it's not safe to do since it can be a third-party thenable, for example from a package using a polyfill or a different library.

A better/simpler check here (if we can't just `await value`) is to do `typeof value.then === 'function'` - it isn't 100% safe (then can be a getter for example) but it's as safe as we can get.

Reply:
This still meeds fixing :)

Reply:
It should be possible to simply use a `Promise.race()` for this

Reply:
fixed